rfcs/01780_policy_delegation.md

```diff
@@ -0,0 +1,308 @@
+
+For instance, let's assume there are three Wasm modules, with hashes
+respectively w_0, w_1, w_2. Also there are two (public, secret) key pairs (p_0,
+s_0), (p_1, s_1).
+
+Let's assume that the modules are supposed to implement some declassification
+logic, e.g. counting the number of people in an image given raw pixels. s_0 and
+s_1 are two parties that verify such models (from correctness, privacy and
+security point of view) and sign them if they (respectively) believe that they
+are trustworthy.
+
+The owner of s_0 signs w_0 and w_1 with s_0, while the owner of s_1 signs w_1
+and w_2 with s_1:
+
+```text
+    ┌ w_0
+s_0 │
+    └ w_1 ┐
+          │ s_1
+      w_2 ┘
+```
+
+## Delegation Manifest
+
+The main contribution of this RFC is to specify a mechanism by which parties may
+publish a list of hashes in a standard format and serve it statically over
+HTTPS.
+
+In the example above, the owner of s_0 would need to publish the signed module
+hashes somewhere, and a client would have to know the value of s_0 itself (the
+actual key used for signing the module), which introduces extra complexity and
+parties around the deployment and publishing of new modules.
+
+Here we introduce a delegation mechanism that is based on the publication of a
+manifest file on an HTTPS endpoint.
+
+A Delegation Manifest is identified (and located) by an HTTPS URL pointing to a
+JSON file with the schema described below.
+
+For instance, Google may publish a delegation manifest under
+`https://google.com/oak/manifests/count_people_from_pixels.json` with the
+following contents:
+
+```json
+{
+  "trusted": [
+    {
+      "type": "manifest",
+      "value": "https://eff.com/oak/manifests/count_people_from_pixels.json"
+    },
+    {
+      "type": "wasm_module",
+      "value": "65c296d60c077db03c7ec90c60107a52221c14bb"
+    },
+    {
+      "type": "wasm_module",
+      "value": "cf26b96746311908f8fe40340a8b6c1f06f5a1b5"
+    }
+  ]
+}
+```
+
+- ~~The `description` field contains a human-readable description of the entries
+  that the manifest refers to.~~ This was considered but then removed, so that
+  it may not be misused by showing it to users in the wrong context.
+- The `trusted` field contains a list of object that are trusted by the
+  publisher to conform to the abstract policy identified by the manifest. Each
+  object in the `trusted` list is composed of the following fields:
+  - `type`: the category of the object (each list of trusted object may have
+    objects of various types); some example values of the `type` fields are:
+    - `manifest`
+    - `wasm_module`
+    - `intel_sgx_mrenclave`
+  - `value`: a string representing a value that uniquely identifies the trusted
+    object; commonly this would be a hash of the object, hex or base64 encoded.
+
+It is possible to include / chain manifests by including the URL of a target
+manifest as a trusted object itself, in which case the chain of URLs is
+recursively expanded at runtime by Oak.
```

Comment:
Expanding at runtime seems risky (e.g. cycles, broken links etc). Perhaps it would be better to pre-download and validate all of this before starting the runtime.

Reply:
I agree, I thought about those issues too, which is why I clarified below that perhaps these should be part of the manifest file and expanded at Oak Runtime startup (when loading the application). Then perhaps periodically refreshed later for revocations (TBD).